File: C++/19.删除链表的倒数第-n-个结点.cpp
/*
 * @lc app=leetcode.cn id=19 lang=cpp
 *
 * [19] 删除链表的倒数第 N 个结点
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if(head == NULL) return head;
        ListNode* fast = head, *low = head;
        while(fast != NULL & n != 0) {
            fast = fast->next;
            n--;
        }
        if(fast == NULL) {
            return head->next;
        }

        while(fast->next != NULL) {
            low = low->next;
            fast = fast->next;
        }
        low->next = low->next->next;

        return head;
    }
};
// @lc code=end

File: Go/19.删除链表的倒数第-n-个结点.go
/*
 * @lc app=leetcode.cn id=19 lang=golang
 *
 * [19] 删除链表的倒数第 N 个结点
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	fast, low := head, head
	for fast != nil && n != 0 {
		fast = fast.Next
		n--
	}
	if fast == nil {
		return head.Next
	}
	for fast.Next != nil {
		low = low.Next
		fast = fast.Next
	}
	low.Next = low.Next.Next
	return head
}

// @lc code=end

